engine/language_client_cffi/src/lib.rs

```diff
@@ -279,3 +308,96 @@ fn call_function_stream_from_c_inner(
 fn on_event(id: u32, result: FunctionResult) {
     safe_trigger_callback(id, true, Ok(result));
 }
+
+#[no_mangle]
+pub extern "C" fn call_collector_function(
+    object: *const libc::c_void,
+    object_type: *const c_char,
+    function_name: *const c_char,
+) -> *const libc::c_void {
+    match call_collector_function_inner(object, object_type, function_name) {
+        Ok(result) => result,
+        Err(e) => {
+            Box::into_raw(Box::new(CString::new(e.to_string()).unwrap())) as *const libc::c_void
+        }
+    }
+}
+
+fn call_collector_function_inner(
+    object: *const libc::c_void,
+    object_type: *const c_char,
+    function_name: *const c_char,
+) -> Result<*const libc::c_void> {
+    let object_type = match unsafe { CStr::from_ptr(object_type) }.to_str() {
+        Ok(s) => s.to_owned(),
+        Err(_) => {
+            return Err(anyhow::anyhow!("Failed to convert object type to string"));
+        }
+    };
+
+    let function_name = match unsafe { CStr::from_ptr(function_name) }.to_str() {
+        Ok(s) => s.to_owned(),
+        Err(_) => {
+            return Err(anyhow::anyhow!("Failed to convert function name to string"));
+        }
+    };
+
+    if object.is_null() {
+        return match (object_type.as_str(), function_name.as_str()) {
+            ("collector", "new") => {
+                let collector = Collector::new(None);
+                Ok(Arc::into_raw(Arc::new(collector)) as *const libc::c_void)
+            }
+            _ => Err(anyhow::anyhow!(
+                "Failed to call collector function: {}",
+                function_name
+            )),
+        };
+    }
+
+    match object_type.as_str() {
+        "collector" => {
+            let collector = unsafe { Arc::from_raw(object as *const Collector) };
+            let clone = collector.clone();
+            let _ = Arc::into_raw(clone);
+
+            match function_name.as_str() {
+                "destroy" => {
+                    let _ = collector.deref();
+                    Ok(null())
+                }
+                "usage" => {
+                    let usage = collector.usage();
+                    Ok(Box::into_raw(Box::new(usage)) as *const libc::c_void)
+                }
+                _ => Err(anyhow::anyhow!(
+                    "Failed to call function: {} on object type: {}",
+                    function_name,
+                    object_type
+                )),
+            }
+        }
+        "usage" => {
+            let usage = unsafe { Box::from_raw(object as *mut Usage) };
```

Comment:
So this has the same problem as Arc where it is now 'owned' by this Box, and gets dropped when the Box is destroyed, which I need to figure out how to handle.